tempesta_fw/t/unit/tfw_bomber.c

```diff
@@ -0,0 +1,496 @@
+				tfw_bomber_desc[i][k].sk->sk_user_data = NULL;
+				ss_close(tfw_bomber_desc[i][k].sk);
+				local_bh_enable();
+				tfw_bomber_desc[i][k].sk = NULL;
+			}
+		}
+	}
+}
+
+static int
+tfw_bomber_thread_connect(void *data)
+{
+	int i, nconnects = 0;
+	int threadn = (int)(long)data;
+	int descidx = threadn * tfw_connects;
+
+	SS_DBG("connect_thread_%02d started\n", threadn);
+	for (i = 0; i < tfw_connects; i++) {
+		if (tfw_bomber_connect(descidx + i) == 0) {
+			nconnects++;
+		}
+	}
+	tfw_bomber_connect_task[threadn] = NULL;
+	atomic_dec(&tfw_bomber_nthread);
+	wake_up(&tfw_bomber_connect_wq);
+	SS_DBG("Thread %d has initiated %d connects out of %d\n",
+	       threadn, nconnects, tfw_connects);
+	return 0;
+}
+
+static void
+tfw_bomber_stop_threads(void)
+{
+	int i;
+
+	for (i = 0; i < tfw_threads; i++) {
+		if (tfw_bomber_connect_task[i]) {
+			kthread_stop(tfw_bomber_connect_task[i]);
+			tfw_bomber_connect_task[i] = NULL;
+		}
+	}
+
+	if (tfw_bomber_finish_task) {
+		kthread_stop(tfw_bomber_finish_task);
+		tfw_bomber_finish_task = NULL;
+	}
+	tfw_bomber_release_sockets();
+}
+
+static int
+tfw_bomber_thread_finish(void *data)
+{
+	int nattempt = atomic_read(&tfw_bomber_connect_nattempt);
+	uint64_t time_max = (uint64_t)get_seconds() + TFW_BOMBER_WAIT_MAX;
+	int ret = 0;
+	int nerror, ncomplete;
+	struct timeval tvs, tve;
+	int delta;
+
+	do_gettimeofday(&tvs);
+
+	set_freezable();
+	do {
+		nerror = atomic_read(&tfw_bomber_connect_nerror);
+		ncomplete = atomic_read(&tfw_bomber_connect_ncomplete);
+
+		if (ncomplete + nerror == nattempt) {
+			break;
+		}
+		wait_event_freezable_timeout(tfw_bomber_finish_wq,
+				kthread_should_stop(),
+				TFW_BOMBER_WAIT_INTVL);
+		if ((uint64_t)get_seconds() > time_max) {
+			SS_ERR("%s exceeded maximum wait time of %d seconds\n",
+			       "tfw_bomber_thread_finish", TFW_BOMBER_WAIT_MAX);
+			break;
+		}
+	} while (!kthread_should_stop() && ncomplete + nerror < nattempt);
+
+	tfw_bomber_send_msgs();
+	tfw_bomber_release_sockets();
+	tfw_bomber_finish_task = NULL;
+
+	do_gettimeofday(&tve);
+
+	bomber_report();
+
+	delta = (tve.tv_sec - tvs.tv_sec) * 1e6 + (tve.tv_usec - tvs.tv_usec);
+	printk("Total time: %d usec\n",delta);
+
+	return ret;
+}
+
+static int
+tfw_bomber_create_tasks(void)
+{
+	int i, ret = 0;
+	struct task_struct *task;
+
+	task = kthread_create(tfw_bomber_thread_finish, 0,
+			      "tfw_bomber_thread_finish");
+	if (IS_ERR_OR_NULL(task)) {
+		ret = PTR_ERR(task);
+		SS_ERR("Unable to create thread: %s (%d)\n",
+		       "tfw_bomber_finish_task", ret);
+		return ret;
+	}
+	tfw_bomber_finish_task = task;
+
+	for (i = 0; i < tfw_threads; i++) {
+		task = kthread_create(tfw_bomber_thread_connect, (void *)(long)i,
+				      "tfw_bomber_thread_connect_%02d", i);
+		if (IS_ERR_OR_NULL(task)) {
+			ret = PTR_ERR(task);
+			SS_ERR("Unable to create a thread: %s%02d (%d)\n",
+			       "tfw_bomber_thread_connect", i, ret);
+			break;
+		}
+		tfw_bomber_connect_task[i] = task;
+	}
```

Comment:
The design is simply wrong: ```tfw_threads``` threads just do reconnects in ```tfw_bomber_thread_connect``` while the only one thread ```tfw_bomber_thread_finish``` sends messages. Each working thread must establish ```tfw_connects``` and send ```tfw_messages``` through each of them. When ```tfw_messages``` are sent through a connection, the connection must be reestablished by the working thread. Thus we need one more argument - ```tfw_total_messages``` which defines number of iterations in a thead as ```iter = tfw_total_messages / tfw_threads / tfw_connects / tfw_messages```.